src/components/vm/consoles/consoles.jsx

```diff
@@ -17,143 +17,250 @@
-            return { serial: newSerial };
+export class ConsoleCardState extends StateObject {
+    constructor () {
+        super();
+        this.type = null;
+        this.vncState = new ConsoleState();
+        this.serialStates = new SerialStates();
 
-        return null;
+        this.follow(this.vncState);
+        this.follow(this.serialStates);
     }
 
-    getDefaultConsole () {
-        const { vm } = this.props;
+    close() {
+        this.vncState.close();
+        this.serialStates.close();
+    }
 
-        if (vm.displays) {
-            if (vm.displays.find(display => display.type == "vnc")) {
-                return 'VncConsole';
-            }
-            if (vm.displays.find(display => display.type == "spice")) {
-                return 'DesktopViewer';
-            }
-        }
+    setType(val) {
+        this.type = val;
+        this.update();
+    }
+}
 
-        const serialConsoleCommand = domainSerialConsoleCommand({ vm });
-        if (serialConsoleCommand) {
-            return 'SerialConsole';
-        }
+export const ConsoleCard = ({ state, vm, config, onAddErrorNotification, isExpanded }) => {
+    const serials = vm.displays && vm.displays.filter(display => display.type == 'pty');
+    const inactive_serials = vm.inactiveXML.displays && vm.inactiveXML.displays.filter(display => display.type == 'pty');
+    const vnc = vm.displays && vm.displays.find(display => display.type == 'vnc');
+    const inactive_vnc = vm.inactiveXML.displays && vm.inactiveXML.displays.find(display => display.type == 'vnc');
+    const spice = vm.displays && vm.displays.find(display => display.type == 'spice');
+    const inactive_spice = vm.inactiveXML.displays && vm.inactiveXML.displays.find(display => display.type == 'spice');
 
-        // no console defined
-        return null;
+    let type = state.type;
+    if (!type) {
+        if (vnc || serials.length == 0)
+            type = "graphical";
+        else
+            type = "text0";
     }
 
-    onDesktopConsoleDownload (type) {
-        const { vm } = this.props;
-        // fire download of the .vv file
-        const consoleDetail = vm.displays.find(display => display.type == type);
+    const actions = [];
+    const tabs = [];
+    let body = null;
+    let body_state = null;
 
-        let address;
-        if (cockpit.transport.host == "localhost") {
-            const app = cockpit.transport.application();
-            if (app.startsWith("cockpit+=")) {
-                address = app.substr(9);
+    tabs.push(<ToggleGroupItem
+                  key="graphical"
+                  text={_("Graphical")}
+                  isSelected={type == "graphical"}
+                  onChange={() => state.setType("graphical")} />);
+
+    if (type == "graphical") {
+        if (vm.state != "running") {
+            if (!inactive_vnc && !inactive_spice) {
+                body = <VncMissing vm={vm} />;
+            } else if (inactive_vnc) {
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2008-6fca6491-20250714-153804-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)